pkg/latte/intent_reauthenticate.go

```diff
@@ -0,0 +1,88 @@
+package latte
+
+import (
+	"context"
+
+	"github.com/authgear/authgear-server/pkg/api/apierrors"
+	"github.com/authgear/authgear-server/pkg/lib/authn/authenticator"
+	"github.com/authgear/authgear-server/pkg/lib/session"
+	"github.com/authgear/authgear-server/pkg/lib/workflow"
+	"github.com/authgear/authgear-server/pkg/util/validation"
+)
+
+func init() {
+	workflow.RegisterPublicIntent(&IntentReauthenticate{})
+}
+
+var IntentReauthenticateSchema = validation.NewSimpleSchema(`
+	{
+		"type": "object",
+		"additionalProperties": false
+	}
+`)
+
+type IntentReauthenticate struct {
+}
+
+func (*IntentReauthenticate) Kind() string {
+	return "latte.IntentReauthenticate"
+}
+
+func (*IntentReauthenticate) JSONSchema() *validation.SimpleSchema {
+	return IntentReauthenticateSchema
+}
+
+func (*IntentReauthenticate) CanReactTo(ctx context.Context, deps *workflow.Dependencies, w *workflow.Workflow) ([]workflow.Input, error) {
+	switch len(w.Nodes) {
+	case 0:
+		return nil, nil
+	case 1:
+		return nil, nil
+	}
+
+	return nil, workflow.ErrEOF
+}
+
+func (i *IntentReauthenticate) ReactTo(ctx context.Context, deps *workflow.Dependencies, w *workflow.Workflow, input workflow.Input) (*workflow.Node, error) {
+	userID, err := i.userID(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	switch len(w.Nodes) {
+	case 0:
+		if err != nil {
+			return nil, err
+		}
+		return workflow.NewSubWorkflow(&IntentAuthenticatePassword{
+			UserID:            userID,
+			AuthenticatorKind: authenticator.KindPrimary,
+		}), nil
+	case 1:
+		return workflow.NewSubWorkflow(&IntentCreateSession{
```

Comment:
IntentCreateSession actually does not implement the logic for reauthenticate. Please refer to do_ensure_session.go to add the missing logic.

Reply:
@louischan-oursky I am not sure if I understand it correctly:
In `do_ensure_session`, if mode is `EnsureSessionModeNoop`, basically nothing will be done except updating the login time.
And the condition of using `EnsureSessionModeNoop` is `SuppressIDPSessionCookie`, which means `x_sso_enabled` is not true.
In latte case, `x_sso_enabled` should be always false, therefore actually we only need to update the login time?

Reply:
In fact, it seems `IntentCreateSession` already handles reauth correctly, as it will skip session creation using `SkipCreate` by checking `SuppressIDPSessionCookie`.

Reply:
Oh, you're correct. I am sorry that I overlooked in latte case SuppressIDPSessionCookie is always true.

Reply:
I refactored the code in latte so that it looks more similar to the original workflow in 20fee68

Reply:
Thank you